### PR TITLE
FlexboxGrid - a lightweight* multiboard system.

### PR DESCRIPTION
**Pull request by Trokkin**
`* - (from user viewpoint)`

Documentation coming soon.

### CODE REVIEW

File: wurst/util/Flexboxgrid.wurst [hunk 1]
```diff
@@ -0,0 +1,380 @@
+package Flexboxgrid
+import LinkedList
+import HashMap
+
+public class FlexboxBuilder
+	Flexboxgrid fbg = new Flexboxgrid()
```
Comment (Frotty) [inline wurst/util/Flexboxgrid.wurst]: type inference

File: wurst/util/Flexboxgrid.wurst [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+
+	protected function getItem(int row, int count) returns multiboarditem
+		return mb.getItem(row, count)
+
+	function setWidth(real width)
+		if width > 0.001
+			this.width = width
+
+	function initialize()
+		mb = CreateMultiboard()
```
Comment (Frotty) [inline wurst/util/Flexboxgrid.wurst]: merge into member?

File: wurst/util/Flexboxgrid.wurst [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+		..display(false)
+
+	/** This function starts a recursive multiboard update */
+	function build()
+		rowItemCount.flush()
+		firstColumn.calculate()
+		mb
```
Comment (Frotty) [inline wurst/util/Flexboxgrid.wurst]: the first call should be in same line. don't just have "mb" floating here

File: wurst/util/Flexboxgrid.wurst [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+
+	function update()
+		firstColumn.update()
+
+	ondestroy
+		destroy rowItemCount
+
+class FlexRow
+	let cols = new LinkedList<FlexCol>
+	int height
```
Comment (Frotty) [inline wurst/util/Flexboxgrid.wurst]: make these types inferred by assigning default values.

File: wurst/util/Flexboxgrid.wurst [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+		height = 0
+		width = 0
+		for row in rows
+			row.calculate()
+			if row.width > width
+				width = row.width
+			height += row.height
+
+	function build(Flexboxgrid fbg, int y, int h, real w2r)
+		int py = y
```
Comment (Frotty) [inline wurst/util/Flexboxgrid.wurst]: type inference

File: wurst/util/Flexboxgrid.wurst [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+	function update(Cell c) returns bool
+
+public enum CellMode
+	None
+	Icon
+	Value
+	Both
+
+public class Cell
+	string value = ""
```
Comment (Frotty) [inline wurst/util/Flexboxgrid.wurst]: type inference

File: wurst/util/Flexboxgrid.wurst [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+
+public interface ProgressBarUpdater
+	/** Returns whether the cell is needed to be updated */
+	function update(CellProgressbar c) returns bool
+
+public class CellProgressbar extends Cell
+	int bars
+	real progress
+
+	colorA bg = COLOR_WHITE
```
Comment (Frotty) [inline wurst/util/Flexboxgrid.wurst]: type inference

File: wurst/util/Flexboxgrid.wurst [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+			destroy row
+		destroy rows
+
+
+public interface CellUpdater
+	/** Returns whether the cell is needed to be updated */
+	function update(Cell c) returns bool
+
+public enum CellMode
+	None
```
Comment (Frotty) [inline wurst/util/Flexboxgrid.wurst]: should be all uppercase

File: wurst/util/Flexboxgrid.wurst [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+		for row in rows
+			row.build(fbg, py, width, w2r)
+			py += row.height
+
+		while py < y + h
+			fbg.getRowNextItem(py)
+			..setStyle(true, false)
+			..setWidth(width * w2r - 0.001)
+			..setValue("")
+			// ..setValue((width * w2r).toString())
```
Comment (Frotty) [inline wurst/util/Flexboxgrid.wurst]: ?!

**Comment by Frotty** [comment]
> Documentation coming soon.

I guess by this you also mean hotdoc for the public functions?

**Comment by Trokkin** [comment]
I meant mostly only the hotdoc for the public functions :D
Also some examples and a short general description you will definitely criticize.

**Comment by Frotty** [comment]
Oh I thought you ment the doc page on the website 😄 

Also, there is still one comment remaining.
Do you have any reason for not doing:
```wurst
public class Flexboxgrid
	var mb = CreateMultiboard()..display(false)
```
❓ 

Another thing on my mind with this is that it's not independent from multiboard, i.e. I can't use the flexboxgrid as a layout anywhere else. I think a cleaner way would be to have it decoupled from multiboard and allow multiple backend implementations via an interface. Use case could be a full screen UI for example.

However as a first step towards an easy to use multiboard lib with closure updaters for the stdlib this be nice enough 😸

**Comment by Trokkin** [comment]
Hm, partially you're right, but using the same interface that multiboard has for an abstract fullscreen UI is quite limiting to be thought of, IMHO.

**Comment by Frotty** [comment]
Only as part of one, of course. E.g. a table for inventory, one for the shop keeper's inventory, 1 cast bar.
These elements should all use some common layout imho. But yeah, find for now.
Btw you can resolve conversations when youve done them,

**Comment by Krypt0n** [comment]
Isn't that a dirty and oldschool usecase for wc3? I think custom UI stuff should be made by lightning or Ímage, that is what I use normally for UI, where is the benefit in missusing multiboard for that?

**Comment by Frotty** [comment]
You misunderstood. No1 wants to use a multiboard for full screen ui, its just about the layout, which could be independent from multiboard and applied to various different backends.

**Comment by Frotty** [comment]
Hello?

**Comment by Trokkin** [comment]
image
Here's an example of what could be done with little effort.

**Comment by Frotty** [comment]
I would prefer a new PR, as this one already loads slowly from comments.
Also update from master for the new ci config.

File: wurst/util/Flexboxgrid.wurst [hunk 10]
```diff
@@ -0,0 +1,366 @@
+package Flexboxgrid
+import LinkedList
+import HashMap
+
+public class FlexboxgridBuilder
+	Container c = new Col() // invariant: c != null
```
Comment (Frotty) [inline wurst/util/Flexboxgrid.wurst]: remove comment and use val.

File: wurst/util/Flexboxgrid.wurst [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+function multiboarditem.preset(real width, string value, string icon, colorA clr)
+	this
+	..setStyle(value != "", icon != "")
+	..setWidth(width - 0.001)
+	..setValue(value)
+	..setIcon(icon)
+	..setValueColor(clr)
+	.release()
+
+abstract class Flex // hello, ricardo
```
Comment (Frotty) [inline wurst/util/Flexboxgrid.wurst]: what is a Flex

File: wurst/util/Flexboxgrid.wurst [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+		var py = y
+		for o in objs
+			o.draw(py, width, o.height_)
+			py += o.height_
+
+		while py < y + height
+			cfbx.nextRowItem(py).preset(width, "", "", COLOR_WHITE)
+			py++
+
+class Table extends Container
```
Comment (Frotty) [inline wurst/util/Flexboxgrid.wurst]: Table class already exists from package Table, rather rename to avoid confusion.

**Comment by Trokkin** [comment]
Does not compile yet -- awaits #241 to be merged.

**Comment by Cokemonkey11** [comment]
@trokkin #241 is merged

**Comment by Frotty** [comment]
Closing cuz of inactivity, missing doc and iirc @Trokkin doesn't use wurst anymore so I don't see this getting accepted.
Thanks again for your efforts 👍

**Pull request closed without merging**